### ISSUE TITLE
[Paid Development]: Only allow social logins-- Disable the ability to enter password on erpnext login page

### ISSUE THREAD

**Issue by adityaduggal** [frappe/erpnext#2291]
There should be an option to disable login with password so that admins can control that certain users can only have social logins. Benefits of such thing would be that the 2-step verification which is prevalent these days could be used on erpnext without fearing for the password being hacked.

**Comment by adityaduggal** [comment]
@rmehta Can I get a quote on this feature also include the following in quote:
 - the feature of login table in this which @anandpdoshi had shown long time back.
 - Login table would show the times of login including their IP address instead of just last login.

**Comment by rmehta** [comment]
@adityaduggal hold a bit, too many issues in the pipeline now.

**Comment by adityaduggal** [comment]
Any plans to quote on this feature we are getting too many phishing emails and also a lot of trojan emails. Can you please help us with this feature.

**Comment by rmehta** [comment]
@adityaduggal what do you mean, via website user?

**Comment by adityaduggal** [comment]
@rmehta I have not used the word Website User....what exactly do you want me to elaborate on?

**Comment by rmehta** [comment]
How exactly do you get trojan emails and phishing stuff?

**Comment by adityaduggal** [comment]
well we are getting it through emails and that is not from or related to erpnext but since we are getting such emails I would like to secure the erpnext system so that even if somebody's password is compromised then we know that the hacker cannot login to erpnext system since all the users would have 2-Factor security on erpnext as well.

**Comment by rmehta** [comment]
This does not help in the case, since the user is already logged in. BTW we have added CSRF token in all POST requests so that there is less chance of someone (https://en.wikipedia.org/wiki/Cross-site_request_forgery) damaging the system due to a malicious email.

**Comment by tundebabzy** [comment]
Is there still interest in the request? It's been open for almost three years without any real progress. Close?

**Comment by revant** [comment]
I managed this with website script.

1. Add a social login (only one for the following script to work)
2. place the following script in `Website Script`

```js
// For Auto login
const loginBtn = $('.btn-social');
if (
  window.location.pathname === "/login" &&
  loginBtn.length > 0
) {
  $('.main-section').hide();
  window.location.href = loginBtn[0].href;
}
```

Make the script better for your needs.

Note: This is just for user experience. Login using Rest API will work `/api/method/login`

**Comment by adityaduggal** [comment]
@revant Thanks for the effort but I think this has lost its relevance since 2FA is now built into frappe and hence I am closing this

**Comment by SilentHunter124** [comment]
Can we reopen this ticket? As far as I understand users can bypass login coming from azure ad by simply choosing their own frappe password. This can lead to serious security incidents as accounts in azure get locked when employee leaves company but he still has his password.... Or is there a workaround?

**Comment by revant** [comment]
when employee leaves, disable the user from ERPNext as well.

**Comment by SilentHunter124** [comment]
Unfortunately this is not iso compliant, a lock needs to be centrally in one place instead of locking the user in 100 different systems and ERPnext…

**Comment by SilentHunter124** [comment]
Idea: Adding an enabled field per each third party provider so it can check if frappe login is denied on login? Is that easy way?

**Comment by revant** [comment]
> Idea: Adding an enabled field per each third party provider so it can check if frappe login is denied on login? Is that easy way?

How will that solve manual login?

Problems: 

- user is disabled from third party auth server.
- user tries to login to ERPNext
- user is denied login on third party end, the user/request never reaches frappe/erpnext to set "denied login"

**Comment by SilentHunter124** [comment]
Okay let‘s imagine:

* User works for company and has access to erpnext, he sets his own password in his profile settings by pressing change password
* User leaves company so instead of logging into companies identity system he selects mail and password as he remembered that he manually set password

with my idea above:

* User is generally not allowed to use mail adress password login method 

right?

**Comment by revant** [comment]
I know you are right! I know that user should be centrally controlled from Authorization Server.

ERPNext is not OAuth 2 app/client that will respect any authorization server. It just provides Social login like wordpress or discourse does.

Again you are correct about the functionality.
How to implement it in code? Any idea about that? When Azure user is disabled, how to set ERPNext user to be disabled?

> Idea: Adding an enabled field per each third party provider so it can check if frappe login is denied on login? Is that easy way?

If that is not set immediately when it is set azure side then your logic about "add a field on doctype and set it on next login" is flawed.

If I know that my user will be disabled if I login using Azure, I'll never login using Azure. I'll just continue with email/username and password

**Comment by SilentHunter124** [comment]
Yeah I know it is a complicated topic and I just search for a solution until somebody does a correct Oauth2 or OIDC Implementation with auto user creation and so on... 

So let's analyze what happens currently:

* User with exact same mail address on azure ad needs to be created on erpnext
* Frappe System assigns it a third party provider called "Frappe" (this is what actually username & password is)
* When this AD User first is logged in by azure, another third party provider called "Office 365" is created also on this user account

This looks like that:

![image](https://user-images.githubusercontent.com/1576808/126039039-27bec5dc-2a57-41d0-a2a1-707c9e444c8f.png)

Normally it is usual in OAuth2 that there is a delay of <TokenLifetime> until the user is logged out from all systems after being disabled. (This is a known flaw of the system, it can only be immediately if there is a webhook). 

BUT: What really needs to be done is closing the backdoor. Even after 1 hour the disabled user still can use the first login method to get into the system. If he is not allowed to do that (by "enabled field" on social provider list) he has to use office365 and there he get's not logged in even if the account is still active in erpnext -> Achieved what wanted... 

Is that understandable? It is only all about the backdoor of using "frappe" social login provider entry on a user account.

EDIT: Also maybe User should not be allowed to create API Keys... hopefully then all backdoors are closed and only azure decides yes or no...

**Comment by revant** [comment]
If you're really dependent on Azure AD add a custom app for Azure AD integration that listens for a webhook. Use `frappe.whitelist()`

In that way users of custom oauth2 server can implement their own back channels with their custom apps.

> until somebody does a correct Oauth2 or OIDC Implementation with auto user creation and so on...

I believe, with programming if we can think precisely what needs to be achieved, it can be achieved.

Also this implies that I've done wrong implementation of Oauth2 and oidc here. Please correct me.

**Comment by SilentHunter124** [comment]
That would be a possibility but webhooks are as far as I remember, not part of OAuth2 Providers. 

Sorry wrong wording, didn‘t mean incorrect, rather fixing that security flaws here and complete the Scenario with User-Creation and permissions by claims from oauth2 provider.

so simple question to you: Do you know an easy way, specific users are only allowed to login by specified range of identity providers?

**Comment by revant** [comment]
> That would be a possibility but webhooks are as far as I remember, not part of OAuth2 Providers.

Yes, my suggestion is to write a separate App for third party integrations. Like "AzureAD Integration", "Google Integration", "Gitlab Integration" and install them as needed. In that way each API can be written as intended by the 3rd party instead of adding complex logic on "Social Login Key" or "User" doctype.

> Do you know an easy way, specific users are only allowed to login by specified range of identity providers?

No easy way.

The callback urls are whitelisted functions and can be overridden by custom app.
Override them to check which provider is logging in and then prevent it? Not an easy way.

**Comment by MohammedNoureldin** [comment]
Hi @SilentHunter124,

did you do or get anything new regarding this?

I got confused and I am not sure what is the current situation now.

So at the moment ERPNext creates an account depending on the token it receives from the IdP and forget about it completely? What if we disabled the user in the IdP, will ERPNext know about this?

It should not necessarily know immediately, this is also acceptable to know about that as soon as the access token expires, but it must know. This makes not much sense, especially in such enterprise app, to login once and then create a completely separate account that knows nothing about what is happening on the side of IdP, where normally all accounts and access permissions are managed.
